packages/ra-core/src/i18n/TranslationProvider.tsx

```diff
@@ -1,102 +1,119 @@
-        const { locale, messages } = props;
-        const polyglot = new Polyglot({
-            locale,
-            phrases: defaultsDeep({ '': '' }, messages, defaultMessages),
-        });
-
-        this.state = {
-            contextValues: {
-                locale,
-                translate: polyglot.t.bind(polyglot),
-            },
-        };
-    }
-
-    componentDidUpdate(prevProps) {
-        if (
-            prevProps.locale !== this.props.locale ||
-            prevProps.messages !== this.props.messages
-        ) {
-            const { locale, messages } = this.props;
+const TranslationProvider: FunctionComponent<Props> = props => {
+    const { i18nProvider, children, locale = 'en' } = props;
+    const { startLoading, stopLoading } = useUpdateLoading();
+    const notify = useNotify();
 
+    const [state, setState] = useSafeSetState({
+        provider: i18nProvider,
+        locale,
+        translate: (() => {
+            const messages = i18nProvider(locale);
+            if (messages instanceof Promise) {
+                throw new Error(
+                    `The i18nProvider returned a Promise for the messages of the default locale (${locale}). Please update your i18nProvider to return the messages of the default locale in a synchronous way.`
+                );
+            }
             const polyglot = new Polyglot({
                 locale,
                 phrases: defaultsDeep({ '': '' }, messages, defaultMessages),
             });
+            return polyglot.t.bind(polyglot);
+        })(),
+    });
 
-            this.setState({
-                contextValues: {
-                    locale,
-                    translate: polyglot.t.bind(polyglot),
-                },
-            });
-        }
-    }
-
-    render() {
-        const { children } = this.props;
-        const { contextValues } = this.state;
+    const setLocale = useCallback(
+        newLocale =>
+            new Promise(resolve => {
+                startLoading();
+                // so we systematically return a Promise for the messages
+                // i18nProvider may return a Promise for language changes,
+                resolve(i18nProvider(newLocale));
+            })
+                .then(messages => {
+                    const polyglot = new Polyglot({
+                        locale: newLocale,
+                        phrases: defaultsDeep(
+                            { '': '' },
+                            messages,
+                            defaultMessages
+                        ),
+                    });
+                    stopLoading();
+                    setState({
+                        provider: i18nProvider,
+                        locale: newLocale,
+                        translate: polyglot.t.bind(polyglot),
+                    });
+                })
+                .catch(error => {
+                    stopLoading();
+                    notify('ra.notification.i18n_error', 'warning');
```

Comment:
This is a technical error. I'm not sure we should notify the end user about it

Reply:
The error occurs while loading (potentially via http) a new set of messages. The problem may be on the user side (networking problem), at this stage there is no way to know. 

I think it's preferable to show an error notification when the language switch fails, otherwise a user who pressed a 'change locale' button may hit the button again and again, thinking their click didn't work.